src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -924,9 +924,65 @@ private static Symbol GetNearestOtherSymbol(ConsList<TypeSymbol> list, TypeSymbo
             Debug.Assert(type.IsInterface);
 
             LookupMembersWithoutInheritance(current, type, name, arity, options, originalBinder, accessThroughType, diagnose, ref useSiteDiagnostics, basesBeingResolved);
-            if ((options & (LookupOptions.NamespaceAliasesOnly | LookupOptions.NamespacesOrTypesOnly)) == 0 && !originalBinder.InCrefButNotParameterOrReturnType)
+            if ((options & LookupOptions.NamespaceAliasesOnly) == 0 && !originalBinder.InCrefButNotParameterOrReturnType)
             {
-                LookupMembersInInterfacesWithoutInheritance(current, type.AllInterfacesWithDefinitionUseSiteDiagnostics(ref useSiteDiagnostics), name, arity, basesBeingResolved, options, originalBinder, accessThroughType, diagnose, ref useSiteDiagnostics);
+                LookupMembersInInterfacesWithoutInheritance(current, GetBaseInterfaces(type, basesBeingResolved, ref useSiteDiagnostics),
+                    name, arity, basesBeingResolved, options, originalBinder, accessThroughType, diagnose, ref useSiteDiagnostics);
+            }
+
+        }
+
+        private static ImmutableArray<NamedTypeSymbol> GetBaseInterfaces(NamedTypeSymbol type, ConsList<TypeSymbol> basesBeingResolved, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            if (basesBeingResolved?.Any() != true)
+            {
+                return type.AllInterfacesWithDefinitionUseSiteDiagnostics(ref useSiteDiagnostics);
+            }
+
+            if (basesBeingResolved.ContainsReference(type.OriginalDefinition))
+            {
+                return ImmutableArray<NamedTypeSymbol>.Empty;
+            }
+
+            var interfaces = type.GetDeclaredInterfaces(basesBeingResolved);
+
+            if (interfaces.IsEmpty)
+            {
+                return ImmutableArray<NamedTypeSymbol>.Empty;
+            }
+
+            // Consumers of the result depend on the sorting performed by AllInterfacesWithDefinitionUseSiteDiagnostics.
+            // Let's use similar sort algorithm.
+            var result = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            var visited = new HashSet<NamedTypeSymbol>(TypeSymbol.EqualsConsiderEverything);
+
+            for (int i = interfaces.Length - 1; i >= 0; i--)
+            {
+                addAllInterfaces(interfaces[i], visited, result, basesBeingResolved);
+            }
+
+            result.ReverseContents();
+
+            foreach (var candidate in result)
+            {
+                candidate.OriginalDefinition.AddUseSiteDiagnostics(ref useSiteDiagnostics);
+            }
+
+            return result.ToImmutableAndFree();
+
+            static void addAllInterfaces(NamedTypeSymbol @interface, HashSet<NamedTypeSymbol> visited, ArrayBuilder<NamedTypeSymbol> result, ConsList<TypeSymbol> basesBeingResolved)
+            {
+                if (@interface.IsInterface && visited.Add(@interface) && !basesBeingResolved.ContainsReference(@interface.OriginalDefinition))
```

Comment:
>visited.Add(@interface) [](start = 46, length = 23)

I think the contract here is that if you add to `visited`, you have committed to adding to `result`.  Otherwise the parts that are produced (taking trimming due to `basesBeingResolved` into account) may be incomplete, leading to an incorrect final result.

One way to fix this would be to swap the last two operands of `&&`.